eZ/Publish/API/Repository/Tests/URLAliasServiceTest.php

```diff
@@ -1600,6 +1600,57 @@ public function testDeleteCorruptedUrlAliases()
         );
     }
 
+    public function testRenamingParentContentDoesntBreakChildAlias(): void
+    {
+        $repository = $this->getRepository();
+        $locationService = $repository->getLocationService();
+        $urlAliasService = $repository->getURLAliasService();
+        $contentTypeService = $repository->getContentTypeService();
+        $contentService = $repository->getContentService();
+
+        $languageCode = 'eng-GB';
+        $contentType = $contentTypeService->loadContentTypeByIdentifier('folder');
+
+        // 1. Create parent folder
+        $folderCreateStruct = $contentService->newContentCreateStruct($contentType, $languageCode);
+        $folderCreateStruct->setField('name', 'A');
+        $folderDraft = $contentService->createContent($folderCreateStruct, [
+            $locationService->newLocationCreateStruct(2),
+        ]);
+        $folder = $contentService->publishVersion($folderDraft->getVersionInfo());
+        $folderLocation = $locationService->loadLocation($folder->contentInfo->getMainLocationId());
+
+        // 2. Create child folder
+        $childCreateStruct = $contentService->newContentCreateStruct($contentType, $languageCode);
+        $childCreateStruct->setField('title', 'B');
+        $childDraft = $contentService->createContent($folderCreateStruct, [
+            $locationService->newLocationCreateStruct($folderLocation->id),
+        ]);
+        $child = $contentService->publishVersion($childDraft->getVersionInfo());
+        $childLocation = $locationService->loadLocation($child->contentInfo->getMainLocationId());
+
+        // 3. Create custom URL alias for child folder
+        $urlAliasService->createUrlAlias($childLocation, '/c/b', $languageCode);
+        $lookup = $urlAliasService->lookup('/c/b');
+
+        self::assertSame('/c/b', $lookup->path);
+
+        // 4. Rename "A" to "C"
+        $folderDraft = $contentService->createContentDraft($folder->contentInfo);
+        $folderUpdateStruct = $contentService->newContentUpdateStruct();
+        $folderUpdateStruct->setField('name', 'C');
+        $renamedFolder = $contentService->updateContent($folderDraft->getVersionInfo(), $folderUpdateStruct);
+        $contentService->publishVersion($renamedFolder->getVersionInfo());
+
+        // Loading aliases shouldn't throw a `BadStateException`
+        $urlAliasService->listLocationAliases($childLocation);
+
+        // Renamed content should have '/c2' path alias
+        $lookupRenamed = $urlAliasService->lookup('C2');
+
+        self::assertSame('/C2', $lookupRenamed->path);
```

Comment:
So now I know what's happening here. The end result is unfortunately the opposite of what the PR describes. A **reusabe** entry is the same entry either used for multiple purposes or that was repurposed. The changes you made here made effectively NOP entries NOT reusable. They're left behind as-is and the secondary entry gets created and named C2 according to the unique naming URL Alias logic.

The flow after the changes:

1. `/a/b` -> normal URL alias, `a` and `b` corresponding content items exist
1. `/c/b` -> custom URL alias, `c` is `NOP` because there's no `c` corresponding content
1. Rename `a` to `c`
1. `/c2` now points to renamed content, `/c` is still a NOP entry therefore it wasn't reused.

What we'd want if we wanted to keep re-usability logic would be:

3. Rename `a` to `c`
4. `/c/b` becomes normal URL alias since `c` corresponds now to proper content node
5. `/a/b` -> is a historical entry redirecting to `/c/b`.

The issue with that is that renaming `a` to `c` causes custom URL alias being overridden and incorrectly cleaned up.
```
 id | parent | text |  action   | alias_redirects | is_alias | is_original | link 
----+--------+------+-----------+-----------------+----------+-------------+------
 43 |     42 | b    | eznode:63 |               0 |        0 |           1 |   43
 45 |     44 | b    | eznode:63 |               0 |        1 |           1 |   45
 42 |      0 | c    | eznode:62 |               1 |        0 |           1 |   42
 46 |      0 | a    | eznode:62 |               0 |        0 |           0 |   42
```
Notice that `id=44` referenced in parent column is missing. It's supposed to be cleaned up via `\eZ\Publish\Core\Persistence\Legacy\Content\UrlAlias\Gateway::cleanupAfterPublish` but that neven happens because it's an edge case not considered before.

----

That being said, maybe your change is good, but what we're doing is actually making NOP not reusable, which is behavior change. Going with the fix for cleanup, even if possible, might lead to another issue - I'm not sure how system is going to behave if custom `/c/b` points to completely different Location. Then making it not reusable would make more sense.

Maybe this is something we can discuss with @adamwojs and @ezsystems/php-dev-team before we decide what direction we should take? Or is it clear for everyone and we can move forward with @barw4's fix and just change the wording?

Reply:
This is obviously a naming mistake, you are correct. I relied too much on that comment above the {{if}} (https://github.com/ezsystems/ezplatform-kernel/pull/385/files#diff-d640f4389daf07a7bd0a474b4181c7b97c2c3208d172269c69fa59151e1fe014L232) and confused myself. My intention was that an alias would get incremented with `uniqueCounter++` and I therefore treated it as 'reusable' 😺. Sorry about that

Reply:
Test improved in https://github.com/ezsystems/ezplatform-kernel/pull/385/commits/bd4088eba026b215debabf3a623d8e64c848828b

Reply:
No problem, URL alias nomenclature  was never easy or straightforward ;)
Now we need to decide if that's the change that we really want. Will discuss this with the Team and probably Adam.